rpc/rpcserver/server.go

```diff
@@ -2346,6 +2346,185 @@ func (s *loaderServer) FetchMissingCFilters(ctx context.Context, req *pb.FetchMi
 	return &pb.FetchMissingCFiltersResponse{}, nil
 }
 
+func (s *loaderServer) RpcSync(req *pb.RpcSyncRequest, svr pb.WalletLoaderService_RpcSyncServer) error {
+	defer zero.Bytes(req.Password)
+
+	// Error if the wallet is already syncing with the network.
+	wallet, walletLoaded := s.loader.LoadedWallet()
+	if walletLoaded {
+		_, err := wallet.NetworkBackend()
+		if err == nil {
+			return status.Errorf(codes.FailedPrecondition, "wallet is loaded and already synchronizing")
+		}
+	}
+
+	if req.DiscoverAccounts && len(req.PrivatePassphrase) == 0 {
+		return status.Errorf(codes.InvalidArgument, "private passphrase is required for discovering accounts")
+	}
+	var lockWallet func()
+	if req.DiscoverAccounts {
+		lock := make(chan time.Time, 1)
+		lockWallet = func() {
+			lock <- time.Time{}
+			zero.Bytes(req.PrivatePassphrase)
+		}
+		defer lockWallet()
+		err := wallet.Unlock(req.PrivatePassphrase, lock)
+		if err != nil {
+			return translateError(err)
+		}
+	}
+
+	s.mu.Lock()
+	chainClient := s.rpcClient
+	s.mu.Unlock()
+
+	// If the rpcClient is already set, you can just use that instead of attempting a new connection.
+	if chainClient == nil {
+		networkAddress, err := cfgutil.NormalizeAddress(req.NetworkAddress,
+			s.activeNet.JSONRPCClientPort)
+		if err != nil {
+			return status.Errorf(codes.InvalidArgument, "Network address is ill-formed: %v", err)
+		}
+		chainClient, err = chain.NewRPCClient(s.activeNet.Params, networkAddress, req.Username,
+			string(req.Password), req.Certificate, len(req.Certificate) == 0)
+		if err != nil {
+			return translateError(err)
+		}
+
+		err = chainClient.Start(svr.Context(), false)
+		if err != nil {
+			if err == rpcclient.ErrInvalidAuth {
+				return status.Errorf(codes.InvalidArgument, "Invalid RPC credentials: %v", err)
+			}
+			return status.Errorf(codes.NotFound, "Connection to RPC server failed: %v", err)
```

Comment:
Should use Unavailable if the server can not be reached, but also this needs to check for a canceled context before assuming that.